drivers/SmartThings/zwave-garage-door-opener/src/ecolink-zw-gdo/init.lua

```diff
@@ -0,0 +1,300 @@
+  device:set_component_to_endpoint_fn(component_to_endpoint)
+  device:set_endpoint_to_component_fn(endpoint_to_component)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY, function(d)
+    device:send(BarrierOperator:Get({}))
+  end)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY*2, function(d)
+    device:send(SensorMultilevel:Get({}))
+  end)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY*3, function(d)
+    device:send(Configuration:BulkGetV2({parameter_offset = 1, number_of_parameters = 6}) )
+  end)
+end
+
+--- Handle Device Added Event
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+local function device_added(driver, device)
+  device:send(BarrierOperator:Get({}))
+  -- Reset contact sensor battery level... This should be pollable on the GDO side.
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.battery.battery(CONTACTSENSOR_BATTERY_LEVEL_NORMAL))
+  -- Reset contact sensor fields
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.tamperAlert.tamper.clear())
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.contactSensor.contact.closed())
+  -- Init barrier door state
+  device:emit_event_for_endpoint(GDO_ENDPOINT_NUMBER,
+    capabilities.doorControl.door.closed())
+end
+
+--- Configuration Report Handler
+---
+--- @param device st.zwave.Device
+local function configure_device_with_updated_config(device)
+  local updated_params = {}
+  updated_params[GDO_CONFIG_PARAM_NO_UNATTENDED_WAIT] = {parameter = device.preferences.closeWaitPeriodSec}
+  updated_params[GDO_CONFIG_PARAM_NO_ACTIVATION_TIME] = {parameter = device.preferences.activationTimeMS}
+  updated_params[GDO_CONFIG_PARAM_NO_OPEN_TIMEOUT] = {parameter = device.preferences.doorOpenTimeoutSec}
+  updated_params[GDO_CONFIG_PARAM_NO_CLOSE_TIMEOUT] = {parameter = device.preferences.doorCloseTimeoutSec}
+  updated_params[GDO_CONFIG_PARAM_NO_SHAKE_SENSE] = {parameter = device.preferences.shakeSensitivity}
+  updated_params[GDO_CONFIG_PARAM_NO_APP_RETRY] = {parameter = device.preferences.applicationLevelRetries}
+  device:send(Configuration:BulkSetV2({
+                                        parameter_offset = 1,
+                                        size = 2,
+                                        handshake = false,
+                                        default = false,
+                                        parameters = updated_params
+                                      }))
+end
+
+--- Handle Device Needs Configuring Event
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+local function do_configure(driver, device)
+  configure_device_with_updated_config(device)
+end
+
+local function device_preferences_updated(driver, device, event, args)
+  -- Did my preference value change
+  configure_device_with_updated_config(device)
+end
```

Comment:
check my math, but I think you could just set both lifecycle handlers to be the function `configure_device...` directly. extra args in lua are just ignored

Reply:
When I try this I get error `.\ecolink-zw-gdo\init.lua:122: attempt to index a nil value (field 'preferences')`
I tried to fix it but I am not sure why it will not work.

Reply:
Scratch that. I was able to fix the issue. `device` needs to be the second parameter of the function, though that leaves an unused parameter `driver.`
`local function configure_device_with_updated_config(driver, device)`